File: JS/Tema4_Ejercicio10.js
/*
 * TEMA 4: ESTRUCTURAS REPETITIVAS - EJERCICIO 10
 * ================================================
 * Generar serie de Fibonacci.
 * 
 * EXPLICACIÓN:
 * - Fibonacci: cada número es suma de los dos anteriores
 * - 0, 1, 1, 2, 3, 5, 8, 13, 21...
 * - Dos métodos: iterativo y recursivo
 */

function fibonacci(n) {
    if (n <= 0) return [];
    if (n === 1) return [0];
    
    const serie = [0, 1];
    
    for (let i = 2; i < n; i++) {
        serie.push(serie[i - 1] + serie[i - 2]);
    }
    
    return serie;
}

function fibonacciRecursivo(n) {
    if (n <= 0) return 0;
    if (n === 1) return 1;
    return fibonacciRecursivo(n - 1) + fibonacciRecursivo(n - 2);
}

console.log("=== SERIE DE FIBONACCI ===");

// Método iterativo
console.log("Primeros 15 números de Fibonacci:");
const serie = fibonacci(15);
console.log(serie.join(", "));

// Método recursivo
console.log("\nUsando recursión:");
const fibRecur = [];
for (let i = 0; i < 10; i++) {
    fibRecur.push(fibonacciRecursivo(i));
}
console.log(fibRecur.join(", "));

// Búsqueda de Fibonacci mayor a 1000
console.log("\nNúmeros de Fibonacci menores a 1000:");
const grandeSerie = [];
let n = 1;
while (true) {
    const num = fibonacciRecursivo(n);
    if (num >= 1000) break;
    grandeSerie.push(num);
    n++;
}
console.log(grandeSerie.join(", "));

/*
 * SALIDA ESPERADA:
 * === SERIE DE FIBONACCI ===
 * Primeros 15 números de Fibonacci:
 * 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377
 * 
 * Usando recursión:
 * 0, 1, 1, 2, 3, 5, 8, 13, 21, 34
 * 
 * Números de Fibonacci menores a 1000:
 * 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987
 */

File: Java/Tema4_Ejercicio10.java
/*
 * TEMA 4: ESTRUCTURAS REPETITIVAS - EJERCICIO 10
 * ================================================
 * Escriba un programa que genere la serie de Fibonacci hasta el término N-ésimo.
 */

import java.util.Scanner;

public class Tema4_Ejercicio10 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("¿Cuántos términos de la serie de Fibonacci deseas? ");
        int n = scanner.nextInt();
        
        System.out.println("\n=== SERIE DE FIBONACCI ===");
        
        if (n <= 0) {
            System.out.println("El número debe ser positivo.");
        } else if (n == 1) {
            System.out.println("1");
        } else {
            long fib1 = 0, fib2 = 1;
            
            for (int i = 1; i <= n; i++) {
                if (i == 1) {
                    System.out.print(fib1);
                } else if (i == 2) {
                    System.out.print(", " + fib2);
                } else {
                    long fib3 = fib1 + fib2;
                    System.out.print(", " + fib3);
                    fib1 = fib2;
                    fib2 = fib3;
                }
            }
            System.out.println();
            
            // Mostrar suma
            System.out.println("\nFibonacci(1) = 0");
            System.out.println("Fibonacci(2) = 1");
            System.out.println("Fibonacci(3) = 1");
            System.out.println("Fibonacci(n) = Fibonacci(n-1) + Fibonacci(n-2)");
        }
        
        scanner.close();
    }
}

/*
 * SALIDA ESPERADA:
 * ¿Cuántos términos de la serie de Fibonacci deseas? 7
 * 
 * === SERIE DE FIBONACCI ===
 * 0, 1, 1, 2, 3, 5, 8
 * 
 * Fibonacci(1) = 0
 * Fibonacci(2) = 1
 * Fibonacci(3) = 1
 * Fibonacci(n) = Fibonacci(n-1) + Fibonacci(n-2)
 */
